cpp/benchmarks/common/random_distribution_factory.hpp

```diff
@@ -62,66 +74,106 @@ double geometric_dist_p(T range_size)
+struct value_generator {
+  using result_type = T;
 
-template <typename T, std::enable_if_t<cuda::std::is_integral_v<T>, T>* = nullptr>
-distribution_fn<T> make_distribution(distribution_id did, T lower_bound, T upper_bound)
-{
-  switch (did) {
-    case distribution_id::NORMAL:
-      return [lower_bound, dist = make_normal_dist(upper_bound - lower_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine) + lower_bound; };
-    case distribution_id::UNIFORM:
-      return [dist = make_uniform_dist(lower_bound, upper_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine); };
-    case distribution_id::GEOMETRIC:
-      return [lower_bound, upper_bound, dist = make_geometric_dist(lower_bound, upper_bound)](
-               std::mt19937& engine) mutable -> T {
-        if (lower_bound <= upper_bound)
-          return dist(engine);
-        else
-          return lower_bound - dist(engine) + lower_bound;
-      };
-    default: CUDF_FAIL("Unsupported probability distribution");
+  value_generator(T lower_bound, T upper_bound, thrust::minstd_rand& engine, Generator gen)
+    : lower_bound(std::min(lower_bound, upper_bound)),
+      upper_bound(std::max(lower_bound, upper_bound)),
+      engine(engine),
+      dist(gen)
+  {
   }
-}
 
-template <typename T, std::enable_if_t<cudf::is_floating_point<T>()>* = nullptr>
+  __device__ T operator()(size_t n)
+  {
+    engine.discard(n);
+    if constexpr (cuda::std::is_integral_v<T> &&
+                  cuda::std::is_floating_point_v<decltype(dist(engine))>) {
+      return std::clamp(static_cast<T>(std::round(dist(engine))), lower_bound, upper_bound);
+    } else {
+      return std::clamp(dist(engine), lower_bound, upper_bound);
+    }
+    // Note: uniform does not need clamp, because already range is guaranteed to be within bounds.
+  }
+
+  T lower_bound;
+  T upper_bound;
+  thrust::minstd_rand engine;
+  Generator dist;
+};
+
+template <typename T>
+using distribution_fn = std::function<rmm::device_uvector<T>(thrust::minstd_rand&, size_t)>;
+
+template <
+  typename T,
+  std::enable_if_t<cuda::std::is_integral_v<T> or cuda::std::is_floating_point_v<T>, T>* = nullptr>
 distribution_fn<T> make_distribution(distribution_id dist_id, T lower_bound, T upper_bound)
 {
   switch (dist_id) {
     case distribution_id::NORMAL:
-      return [lower_bound, dist = make_normal_dist(upper_bound - lower_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine) + lower_bound; };
+      return [lower_bound, upper_bound, dist = make_normal_dist(lower_bound, upper_bound)](
+               thrust::minstd_rand& engine, size_t size) -> rmm::device_uvector<T> {
+        rmm::device_uvector<T> result(size, rmm::cuda_stream_default);
+        thrust::tabulate(thrust::device,
+                         result.begin(),
+                         result.end(),
+                         value_generator{lower_bound, upper_bound, engine, dist});
```

Comment:
Since this file is making device calls, I feel that it should be a `.cuh` instead of a `.hpp`. Unless the `tabulate` calls can be moved into a `.cu` file. Considering that `is_floating_point` is only 2 types (float, double) it is probably possible to do.